File: sdks/go/src/cortex/validate.go
package cortex

import (
	"fmt"
)

func edgeHasKey(edge map[string]any, key string) bool {
	_, ok := edge[key]
	return ok
}

// ValidateIncidentContext checks that an IncidentContext has required fields.
func ValidateIncidentContext(ctx *IncidentContext) error {
	if ctx == nil {
		return fmt.Errorf("cortex: incident context is nil")
	}
	if ctx.IncidentID == "" {
		return fmt.Errorf("cortex: incident_id is required")
	}
	if ctx.Timestamp == "" {
		return fmt.Errorf("cortex: timestamp is required")
	}
	if ctx.Confidence < 0 || ctx.Confidence > 1 {
		return fmt.Errorf("cortex: confidence must be in [0.0, 1.0], got %f", ctx.Confidence)
	}
	return nil
}

// ValidateGraphView checks that a GraphView has valid structure.
func ValidateGraphView(gv *GraphView) error {
	if gv == nil {
		return fmt.Errorf("cortex: graph view is nil")
	}
	for i, node := range gv.Nodes {
		if _, ok := node["id"]; !ok {
			return fmt.Errorf("cortex: node %d missing 'id'", i)
		}
	}
	for i, edge := range gv.Edges {
		hasSource := edgeHasKey(edge, "source") || edgeHasKey(edge, "from_node_id")
		hasTarget := edgeHasKey(edge, "target") || edgeHasKey(edge, "to_node_id")
		if !hasSource {
			return fmt.Errorf("cortex: edge %d missing 'source' or 'from_node_id'", i)
		}
		if !hasTarget {
			return fmt.Errorf("cortex: edge %d missing 'target' or 'to_node_id'", i)
		}
	}
	return nil
}

// ValidateRemediation checks that a Remediation has required fields.
func ValidateRemediation(r *Remediation) error {
	if r == nil {
		return fmt.Errorf("cortex: remediation is nil")
	}
	if r.IncidentID == "" {
		return fmt.Errorf("cortex: incident_id is required")
	}
	if r.Action == "" {
		return fmt.Errorf("cortex: action is required")
	}
	return nil
}

// ValidateRemediationFeedback checks that RemediationFeedback has required fields.
func ValidateRemediationFeedback(rf *RemediationFeedback) error {
	if rf == nil {
		return fmt.Errorf("cortex: remediation feedback is nil")
	}
	if rf.RemediationID == "" {
		return fmt.Errorf("cortex: remediation_id is required")
	}
	if rf.IncidentID == "" {
		return fmt.Errorf("cortex: incident_id is required")
	}
	if rf.Outcome == "" {
		return fmt.Errorf("cortex: outcome is required")
	}
	return nil
}

File: sdks/rs/src/cortex/validate.rs
use super::schema::{GraphView, IncidentContext, Remediation, RemediationFeedback};

/// Validation error for cortex API requests.
#[derive(Clone, Debug)]
pub struct CortexValidationError {
    pub field: String,
    pub message: String,
}

impl std::fmt::Display for CortexValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.field, self.message)
    }
}

impl std::error::Error for CortexValidationError {}

/// Validate an incident context before sending to cortex API.
/// Checks required fields are present and well-formed.
pub fn validate_incident_context(ctx: &IncidentContext) -> Result<(), CortexValidationError> {
    if ctx.incident_id.trim().is_empty() {
        return Err(CortexValidationError {
            field: "incident_id".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    if ctx.timestamp.trim().is_empty() {
        return Err(CortexValidationError {
            field: "timestamp".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    if ctx.confidence < 0.0 || ctx.confidence > 1.0 {
        return Err(CortexValidationError {
            field: "confidence".to_string(),
            message: "must be between 0.0 and 1.0".to_string(),
        });
    }
    Ok(())
}

/// Validate a graph view response from cortex API.
pub fn validate_graph_view(graph: &GraphView) -> Result<(), CortexValidationError> {
    for (i, node) in graph.nodes.iter().enumerate() {
        if node.get("id").is_none() {
            return Err(CortexValidationError {
                field: format!("nodes[{i}].id"),
                message: "each node must have an id".to_string(),
            });
        }
    }
    for (i, edge) in graph.edges.iter().enumerate() {
        if edge.get("source").is_none() || edge.get("target").is_none() {
            return Err(CortexValidationError {
                field: format!("edges[{i}]"),
                message: "each edge must have source and target".to_string(),
            });
        }
    }
    Ok(())
}

/// Validate remediation before recording.
pub fn validate_remediation(remediation: &Remediation) -> Result<(), CortexValidationError> {
    if remediation.incident_id.trim().is_empty() {
        return Err(CortexValidationError {
            field: "incident_id".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    if remediation.action.trim().is_empty() {
        return Err(CortexValidationError {
            field: "action".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    Ok(())
}

/// Validate remediation feedback before recording.
pub fn validate_remediation_feedback(
    feedback: &RemediationFeedback,
) -> Result<(), CortexValidationError> {
    if feedback.remediation_id.trim().is_empty() {
        return Err(CortexValidationError {
            field: "remediation_id".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    if feedback.incident_id.trim().is_empty() {
        return Err(CortexValidationError {
            field: "incident_id".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    if feedback.outcome.trim().is_empty() {
        return Err(CortexValidationError {
            field: "outcome".to_string(),
            message: "must be non-empty".to_string(),
        });
    }
    Ok(())
}
